CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+    }
+
+    record $fields => {
+      if is_bitflags($fields):
+        if $fields.len() > 32:
+          lift(direction, type, [read_8_bytes(ptr)])
+        else:
+          lift(direction, type, [read_4_bytes(ptr)])
+
+      else:
+        fields = []
+        for field in $fields:
+          ptr = align_up_to(ptr, align(direction, field))
+          fields.push(read_from_memory(direction, field, ptr))
+          ptr += size(direction, field)
+        fields as $language_struct
+    }
+
+    variant $cases => {
+      discrim = read_4_bytes(ptr)
+      case = $cases[discrim]
+      ptr = align_up_to(ptr + 4, align(case))
+      payload = read_from_memory(case, ptr)
+
+      (discrim, payload) as $language_enum
+    }
+
+    shared-handle => lift(direction, type, [read_4_bytes(ptr)])
+
+    push-buffer | pull-buffer => {
+        assert(direction == "export")
+        lift(direction, type, [read_4_bytes(ptr)])
+    }
+```
+
+## Lowering values
+
+Lowering is very similar to lifting, it's basically just the opposite in various
+directions.
+
+```
+def lower_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lower(direction, ty, values))
+
+def lower(direction, type, value):
+  match type
+    f32 => [value as wasm_f32]
+    f64 => [value as wasm_f64]
+    s8 => [value as wasm_i32]
+    u8 => [value as wasm_i32]
+    s16 => [value as wasm_i32]
+    u16 => [value as wasm_i32]
+    s32 => [value as wasm_i32]
+    u32 => [value as wasm_i32]
+    s64 => [value as wasm_i64]
+    u64 => [value as wasm_i64]
+    char => [value as wasm_i32]
+
+    # It's the responsibility of the native language to supply a list where the
+    # in-memory representation of the list matches the canonical ABI.
+    list $ty => {
+      [value.ptr as wasm_i32, value.len as wasm_i32]
+    }
+
+    record $fields => {
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              [value.bits as wasm_i64]
+          else:
+              [value.bits as wasm_i32]
+        else:
+	  wasm_vals = []
+	  for field in $fields:
+	    wasm_vals.extend(lower(direction, field, value[field.name]))
+          return wasm_vals
+    }
+
+    variant $cases => {
+      match value {
+        the_case(payload)=> {
+          # Recursively lower the optional payload
+          payload_vals = lower(direction, $cases[the_case], payload)
+
+          # Upcast each individual type in the payload to the shared type for
+          # each arm of the variant
+          payload_types = flatten_cases(direction, $cases)
+          for i in 0..len(payload_vals):
+              payload_vals[i] = upcast(payload_vals[i], payload_types[i])
+
+          # If the shared type has more values than this payload generated we
+          # insert zero values to pad
+          for i in len(payload_vals)..len(payload_types)
+              payload_vals.push(zero_value(payload_types[i]))
+
+          [the_case as wasm_i32] + payload_vals
```

Comment:
```suggestion
          [$cases.index(the_case) as wasm_i32] + payload_vals
```
(My understanding is that `the_case` is a string label, and thus we should be explicit that the integer ordering comes from what is declared in the (local) type.)